pybmix/core/chain.py

```diff
@@ -0,0 +1,122 @@
+        """Extracts the chain relative to 'param_name' in a numpy format
+        If 'param_name' is in a nested field, use the dot syntax to
+        join fields.
+
+        Parameters
+        ----------
+        param_name: string
+            the name of the parameter to extract
+        to_arviz: bool (default False)
+            if True, converts the chain to an instance of
+            arviz.data.inference_data.InferenceData
+
+        Example
+        -------
+            Suppose that the chain is a list of MarginalState messages and 
+            we want to get the cardinality of the first cluster
+        >>> chain = mixture_model.get_chain()
+        >>> card_chain = chain.extract("cluster_states[0].cardinality")
+        """
+        chain = self.chain if self.chain is not None else self.serialized_chain
+        extractor = self._get_extractor(param_name)
+        out = None
+
+        # we need to perform these checks because we're checking also
+        # for base classes
+        if isinstance(extractor(chain[0]), RepeatedScalarContainer):
+            out = self._extract_repeated(chain, extractor)
+        elif isinstance(extractor(chain[0]), matrix_pb2.Vector):
+            out = self._extract_vector(chain, extractor)
+        elif isinstance(extractor(chain[0]), matrix_pb2.Matrix):
+            out = self._extract_matrix(chain, extractor)
+        else:
+            try:
+                out = np.array([extractor(x) for x in chain])
+            except Exception as e:
+                logging.error(e)
+
+        if out is not None and to_arviz:
+            out = self.to_arviz(param_name, chain)
+
+        return out
+    
+    def _extract_repeated(self, chain, extractor):
+        first = extractor(chain[0])
+        out = np.empty((len(chain), len(first)), dtype=type(first[0]))
+        out[0, :] = first
+        for i in range(1, len(chain)):
+            out[i, :] = extractor(chain[i])
+        return out
+
+    def _extract_vector(self, chain, extractor):
+        def to_numpy(msg):
+           return np.array(msg.data) 
+
+        first = extractor(chain[0])
+        out = np.empty((len(chain), first.size))
+        for i in range(len(chain)):
+            out[i, :] = to_numpy(extractor(chain[i]))
+        return out
+
+    def _extract_matrix(self, chain, extractor):
+        def to_numpy(msg):
+            order = "T" if msg.rowmajor else "F"
+            return np.array(msg.data).reshape(msg.rows, msg.cols, order=order)
+        
+        first = extractor(chain[0])
+        out = np.empty((len(chain), first.rows, first.cols))
+        for i in range(len(chain)):
+            out[i, :, :] = to_numpy(extractor(chain[i]))
+        return out
+
+
+    def _get_extractor(self, param_name):
+        if self.chain is None:
+            def extractor(x): return get_field(self._deserialize(x), param_name)
+        else:
+            def extractor(x): return get_field(x, param_name)
+
+        return extractor
```

Comment:
Wouldn't it be better to just initialize the appropriate `extractor` in the constructor and store it as a class member? It seems pointless to pass it to all these member functions

Reply:
But the `extractor` depends on the `param_name`, and we might want to extract several parameters from the same chain.